libs/orders/src/lib/components/order-list-manager/order-list-manager.tsx

```diff
@@ -1,79 +1,145 @@
+  const newRows = useRef(0);
+  const scrolledToTop = useRef(true);
   const variables = useMemo(() => ({ partyId }), [partyId]);
 
-  // Apply updates to the table
-  const update = useCallback(({ delta }: { delta: OrderSub_orders[] }) => {
-    if (!gridRef.current) {
-      return false;
+  const addNewRows = useCallback(() => {
+    if (newRows.current === 0) {
+      return;
     }
-    const incoming = prepareIncomingOrders(delta);
-
-    const updateRows: OrderFields[] = [];
-    const add: OrderFields[] = [];
+    if (totalCountRef.current !== undefined) {
+      totalCountRef.current += newRows.current;
+    }
+    newRows.current = 0;
+    if (!gridRef.current?.api) {
+      return;
+    }
+    gridRef.current.api.refreshInfiniteCache();
+  }, []);
 
-    incoming.forEach((d) => {
+  const update = useCallback(
+    ({
+      data,
+      delta,
+    }: {
+      data: (Orders_party_ordersConnection_edges | null)[];
+      delta: OrderFields[];
+    }) => {
       if (!gridRef.current?.api) {
-        return;
+        return false;
       }
-
-      const rowNode = gridRef.current.api.getRowNode(d.id);
-
-      if (rowNode) {
-        if (!isEqual(d, rowNode.data)) {
-          updateRows.push(d);
+      if (!scrolledToTop.current) {
+        const createdAt = dataRef.current?.[0]?.node.createdAt;
+        if (createdAt) {
+          newRows.current += delta.filter(
+            (trade) => trade.createdAt > createdAt
+          ).length;
         }
-      } else {
-        add.push(d);
       }
-    });
-
-    if (updateRows.length || add.length) {
-      gridRef.current.api.applyTransactionAsync({
-        update: updateRows,
-        add,
-        addIndex: 0,
-      });
-    }
+      dataRef.current = data;
+      gridRef.current.api.refreshInfiniteCache();
+      return true;
+    },
+    []
+  );
 
-    return true;
-  }, []);
+  const insert = useCallback(
+    ({
+      data,
+      totalCount,
+    }: {
+      data: Orders_party_ordersConnection_edges[];
+      totalCount?: number;
+    }) => {
+      dataRef.current = data;
+      totalCountRef.current = totalCount;
+      return true;
+    },
+    []
+  );
 
-  const { data, error, loading } = useDataProvider({
+  const { data, error, loading, load, totalCount } = useDataProvider({
     dataProvider,
     update,
+    insert,
     variables,
   });
+  totalCountRef.current = totalCount;
+  dataRef.current = data;
 
-  const orders = useMemo(() => {
-    if (!data) {
-      return null;
+  const getRows = async ({
+    successCallback,
+    failCallback,
+    startRow,
+    endRow,
+  }: IGetRowsParams) => {
+    startRow += newRows.current;
+    endRow += newRows.current;
+    console.log('getRows');
+    try {
+      if (dataRef.current && dataRef.current.indexOf(null) < endRow) {
+        await load();
+      }
+      const rowsThisBlock = dataRef.current
+        ? dataRef.current.slice(startRow, endRow).map((edge) => edge?.node)
+        : [];
+      let lastRow = -1;
```

Comment:
suggestion (non-blocking): i'd create a `const lastRow = getLastRow(totalCountRef.current)` function here, to clean up a little the large amount of nested if blocks here